crates/biome_js_analyze/src/lint/nursery/no_useless_undefined.rs

```diff
@@ -0,0 +1,447 @@
+use biome_analyze::{
+    context::RuleContext, declare_lint_rule, ActionCategory, Ast, FixKind, Rule, RuleDiagnostic,
+    RuleSource,
+};
+use biome_console::markup;
+use biome_js_factory::make::{self, js_function_body, js_variable_declarator_list};
+use biome_js_syntax::{
+    AnyJsExpression, AnyJsFunctionBody, JsArrayBindingPatternElement, JsArrowFunctionExpression,
+    JsCallExpression, JsFormalParameter, JsLanguage, JsObjectBindingPatternShorthandProperty,
+    JsReturnStatement, JsSyntaxToken, JsVariableDeclarator, JsVariableStatement, JsYieldArgument,
+    T,
+};
+use biome_rowan::{
+    chain_trivia_pieces, declare_node_union, AstNode, AstSeparatedList, BatchMutationExt,
+    SyntaxTriviaPiece, TextRange, TriviaPieceKind,
+};
+
+use crate::JsRuleAction;
+
+declare_lint_rule! {
+    /// Disallow useless `undefined`.
+    ///
+    /// `undefined` is the default value for new variables, parameters, return statements, etc… so specifying it doesn't make any difference.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```js,expect_diagnostic
+    /// let foo = undefined;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const {foo = undefined} = bar;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// const noop = () => undefined;
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// function foo() {
+    ///    return undefined;
+    /// }
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// function* foo() {
+    ///   yield undefined;
+    /// }
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// function foo(bar = undefined) {}
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// function foo({bar = undefined}) {}
+    /// ```
+    ///
+    /// ```js,expect_diagnostic
+    /// foo(undefined);
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```js
+    /// let foo;
+    /// const {foo} = bar;
+    /// function foo() {
+    ///   return;
+    /// }
+    /// function* foo() {
+    ///   yield;
+    /// }
+    /// function foo(bar) {}
+    /// function foo({bar}) {}
+    /// foo();
+    /// ```
+    ///
+    pub NoUselessUndefined {
+        version: "next",
+        name: "noUselessUndefined",
+        language: "js",
+        fix_kind: FixKind::Unsafe,
+        sources: &[RuleSource::EslintUnicorn("no-useless-undefined")],
+        recommended: false,
+    }
+}
+
+declare_node_union! {
+    pub RuleQuery = JsVariableStatement
```

Comment:
We always use `Any*` when joining nodes. In this case it should be called `AnyRuleQuery` or some other name that follows the internal naming convention.